core/json/src/main/scala/net/liftweb/json/JsonParser.scala

```diff
@@ -83,89 +85,118 @@ object JsonParser {
+    var i = 0
+    while (i < 10) {
+      chars(i + '0') = i
+      i += 1
+    }
+    i = 0
+    while (i < 16) {
+      chars(i + 'a') = 10 + i
+      chars(i + 'A') = 10 + i
+      i += 1
+    }
+    chars
+  }
+
+  private[json] def unquote(string: String): String =
     unquote(new JsonParser.Buffer(new java.io.StringReader(string), false))
-  
-  private[json] def unquote(buf: JsonParser.Buffer): String = {
-    def unquote0(buf: JsonParser.Buffer, base: String): String = {
-      val s = new java.lang.StringBuilder(base)
+
+  private[this] def unquote(buf: JsonParser.Buffer): String = {
+    def unquote0(buf: JsonParser.Buffer): String = {
+      val builder = buf.builder
+      builder.delete(0, builder.length())
       var c = '\\'
       while (c != '"') {
         if (c == '\\') {
-          buf.next match {
-            case '"'  => s.append('"')
-            case '\\' => s.append('\\')
-            case '/'  => s.append('/')
-            case 'b'  => s.append('\b')
-            case 'f'  => s.append('\f')
-            case 'n'  => s.append('\n')
-            case 'r'  => s.append('\r')
-            case 't'  => s.append('\t')
+          buf.substring(intoBuilder = true)
+          (buf.next: @switch) match {
+            case '"'  => builder.append('"')
+            case '\\' => builder.append('\\')
+            case '/'  => builder.append('/')
+            case 'b'  => builder.append('\b')
+            case 'f'  => builder.append('\f')
+            case 'n'  => builder.append('\n')
+            case 'r'  => builder.append('\r')
+            case 't'  => builder.append('\t')
             case 'u' =>
+              var byte = 0
+              var finalChar = 0
               val chars = Array(buf.next, buf.next, buf.next, buf.next)
-              val codePoint = Integer.parseInt(new String(chars), 16)
-              s.appendCodePoint(codePoint)
-            case _ => s.append('\\')
+              while (byte < 4) {
+                finalChar = (finalChar << 4) | HexChars(chars(byte).toInt)
+                byte += 1
+              }
+              builder.appendCodePoint(finalChar.toChar)
+            case _ =>
+              builder.append('\\')
           }
-        } else s.append(c)
+          buf.mark
+        }
         c = buf.next
       }
-      s.toString
+      buf.substring(intoBuilder = true)
+      builder.toString
     }
 
     buf.eofIsFailure = true
     buf.mark
     var c = buf.next
+    var forcedReturn: String = null
     while (c != '"') {
-      if (c == '\\') {
-        val s = unquote0(buf, buf.substring)
-        buf.eofIsFailure = false
-        return s
+      (c: @switch) match {
+        case '\\' =>
+          forcedReturn = unquote0(buf)
+          c = '"'
+        case _ =>
+          c = buf.next
       }
-      c = buf.next
     }
     buf.eofIsFailure = false
-    buf.substring
+    forcedReturn match {
+      case null => new String(buf.substring())
+      case _ => forcedReturn
+    }
```

Comment:
I'm inclined to think that an if statement is syntactically prettier in this case, but that could just be me.